### ISSUE TITLE
[java] ExhaustiveSwitchHasDefault improve violation message

### ISSUE THREAD

**Issue by armandino** [pmd/pmd#6984]
**Affects PMD Version:**  7.26.0 (also reproduced on 7.27.0-SNAPSHOT)

**Rule:**  `ExhaustiveSwitchHasDefault` (https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#exhaustiveswitchhasdefault)

**Description:**

The switch below is not exhaustive. `NEW` is only handled by default. Yet PMD reports "exhaustive even without the default case". Removing default produces a compile error (the switch expression does not cover all possible input values).

**Code Sample demonstrating the issue:**

```java
public final class Reproducer {
    enum State { NEW, ACTIVE, DONE, VOID }

    static String apply(State state) {
        return switch (state) {
            case ACTIVE -> "active";
            case DONE -> "done";
            case VOID -> "void";
            case null, default -> throw new IllegalStateException("unsupported: " + state);
        };
    }
}
```

**Expected outcome:** PMD should not report a violation.

**Running PMD through:** Maven (7.26.0), CLI (7.27.0-SNAPSHOT)

**Related:**

 - #5793
 - #5800

**Comment by UncleOwen** [comment]
No, this isn't a false positive. "Exhaustive" in this context doesn't mean "all possible cases are there" but "the compiler can check whether all possible cases are there". See the rule's description.

But I agree that the violation message could be improved.

**Comment by armandino** [comment]
@UncleOwen thank you for clarifying. Couple of options for the error message if there's interest in updating it (happy to submit a PR):

- A compiler-checkable type should omit the default case
- Exhaustive enum/sealed switch loses its safety net once default case is added

**Comment by UncleOwen** [comment]
The first one is very abstract. I like your second suggestion, especially that is mentions that this is about sealed types - but maybe try to make "safety net" more explicit?

PRs are definitely welcome!

**Comment by zbynek** [comment]
Maybe the rule could be rewritten in Java and the error message enhanced to list the enum constants / sealed interface implementations that are needed to replace `default`?

**Comment by armandino** [comment]
Other alternatives:

- Exhaustive enum/sealed switch stops catching missing cases at compile time once default is added
- Exhaustive enum/sealed switch statement stops catching missing cases at compile time once default case is added

**Comment by UncleOwen** [comment]
I was thinking something like

"A default case in a switch on sealed classes or enums prevents the compiler from checking that all cases are present."

@zbynek It's already a java rule in current main :) Could you give an example of how this would look like? Just "(missing case: NEW)" added to whatever the new message is?

**Comment by zbynek** [comment]
I was thinking something like

"Replace default case with explicit cases (NEW) to make sure compiler keeps checking that all cases are present."

**Comment by UncleOwen** [comment]
I like it!

**Comment by armandino** [comment]
Should it truncate the list if a failure has too many values to report?

> "Replace default case with explicit cases (FOO, BAR, BAZ, ...) to make sure compiler keeps checking that all cases are present."

**Comment by UncleOwen** [comment]
Yes!

**Comment by AzazelSensei** [comment]
Opened #6996. The message now lists the missing cases and truncates after three.
